File: boards/micropython-esp32/treadmill.py
import time
from machine import Pin, ADC, PWM, I2C
from mcp23017 import MCP23017, IN, OUT
from config import (
    SPEED_PIN, SPEED_POT_PIN,
    PWM_FREQ, MAX_DUTY_CYCLE,
    SPEED_ANALOG_THRESHOLD, MAX_SPEED_MPH,
    MAX_INCLINE_LEVEL, MIN_INCLINE_LEVEL, INCLINE_BTN_ACTIVE_LOW,
    ENCODER_SW,
    MCP_I2C_SDA, MCP_I2C_SCL, MCP_I2C_ADDR,
    MCP_INCLINE_UP, MCP_INCLINE_DOWN,
    MCP_INCLINE_UP_BTN, MCP_INCLINE_DOWN_BTN, MCP_SAFETY_KEY,
)

_MAX_ADC = 4095   # ESP32 12-bit ADC
_MAX_DUTY = 1023  # MicroPython PWM duty range
_INCLINE_DEBOUNCE_MS = 300
_SW_DEBOUNCE_MS = 50


class Treadmill:
    def __init__(self):
        self.speed_pwm = PWM(Pin(SPEED_PIN), freq=PWM_FREQ, duty=0)

        # MCP23017 on a dedicated I2C bus carries incline relays, incline
        # buttons, and the safety key.  Speed PWM stays on a native GPIO.
        i2c = I2C(1, sda=Pin(MCP_I2C_SDA), scl=Pin(MCP_I2C_SCL), freq=400000)
        self.mcp = MCP23017(i2c, MCP_I2C_ADDR)
        self.mcp.pin_mode(MCP_INCLINE_UP, OUT)
        self.mcp.pin_mode(MCP_INCLINE_DOWN, OUT)
        self.mcp.output(MCP_INCLINE_UP, 0)
        self.mcp.output(MCP_INCLINE_DOWN, 0)
        # Buttons use the expander's internal pull-ups when active-low.
        self.mcp.pin_mode(MCP_INCLINE_UP_BTN, IN, pull_up=INCLINE_BTN_ACTIVE_LOW)
        self.mcp.pin_mode(MCP_INCLINE_DOWN_BTN, IN, pull_up=INCLINE_BTN_ACTIVE_LOW)

        self.speed_pot = ADC(Pin(SPEED_POT_PIN))
        self.speed_pot.atten(ADC.ATTN_11DB)  # full 0–3.3V range

        self.safety_key = MCP_SAFETY_KEY
        if self.safety_key is not None:
            self.mcp.pin_mode(self.safety_key, IN, pull_up=True)
        self.encoder_sw = Pin(ENCODER_SW, Pin.IN, Pin.PULL_UP)

        self.safety_triggered = False
        self.incline_level = 0
        self.speed_mph = 0.0
        self.elapsed_seconds = 0
        self.running = False

        self._last_btn_up = False
        self._last_btn_down = False
        self._last_incline_ms = 0
        self._last_sw = False
        self._last_sw_ms = 0
        self._session_start = None

    def update(self):
        pot_value = self.speed_pot.read()
        safety_active = self.safety_key is not None and self.mcp.input(self.safety_key) == 0

        if safety_active and not self.safety_triggered:
            self._stop_all()
            self.safety_triggered = True
            self.running = False
            return

        if self.safety_triggered:
            self._stop_all()
            if pot_value < SPEED_ANALOG_THRESHOLD:
                self.safety_triggered = False
            return

        self._check_encoder_click()

        if not self.running:
            self._stop_all()
            return

        self._update_incline()
        self._update_speed(pot_value)

        if self.speed_mph > 0 and self._session_start is None:
            self._session_start = time.time()

        if self._session_start is not None:
            self.elapsed_seconds = time.time() - self._session_start

    def _check_encoder_click(self):
        now = time.ticks_ms()
        pressed = self.encoder_sw.value() == 0  # active-low
        if pressed and not self._last_sw and time.ticks_diff(now, self._last_sw_ms) > _SW_DEBOUNCE_MS:
            self.running = not self.running
            self._last_sw_ms = now
        self._last_sw = pressed

    def _btn_pressed(self, mcp_pin):
        val = self.mcp.input(mcp_pin)
        return val == 0 if INCLINE_BTN_ACTIVE_LOW else val == 1

    def _update_incline(self):
        now = time.ticks_ms()
        debounce_ok = time.ticks_diff(now, self._last_incline_ms) > _INCLINE_DEBOUNCE_MS

        up = self._btn_pressed(MCP_INCLINE_UP_BTN)
        down = self._btn_pressed(MCP_INCLINE_DOWN_BTN)

        if up and not down:
            self.mcp.output(MCP_INCLINE_UP, 1)
            self.mcp.output(MCP_INCLINE_DOWN, 0)
            if not self._last_btn_up and debounce_ok:
                self.incline_level = min(self.incline_level + 1, MAX_INCLINE_LEVEL)
                self._last_incline_ms = now
        elif down and not up:
            self.mcp.output(MCP_INCLINE_UP, 0)
            self.mcp.output(MCP_INCLINE_DOWN, 1)
            if not self._last_btn_down and debounce_ok:
                self.incline_level = max(self.incline_level - 1, MIN_INCLINE_LEVEL)
                self._last_incline_ms = now
        else:
            self.mcp.output(MCP_INCLINE_UP, 0)
            self.mcp.output(MCP_INCLINE_DOWN, 0)

        self._last_btn_up = up
        self._last_btn_down = down

    def _update_speed(self, pot_value):
        if pot_value <= SPEED_ANALOG_THRESHOLD:
            self.speed_pwm.duty(0)
            self.speed_mph = 0.0
            return

        duty = int((pot_value / _MAX_ADC) * MAX_DUTY_CYCLE * _MAX_DUTY)
        duty = min(duty, int(MAX_DUTY_CYCLE * _MAX_DUTY))
        self.speed_pwm.duty(duty)
        self.speed_mph = (duty / (_MAX_DUTY * MAX_DUTY_CYCLE)) * MAX_SPEED_MPH

    def _stop_all(self):
        self.speed_pwm.duty(0)
        self.mcp.output(MCP_INCLINE_UP, 0)
        self.mcp.output(MCP_INCLINE_DOWN, 0)
        self.speed_mph = 0.0

File: boards/Tradmill/treadmill.cpp
#include "treadmill.h"
#include "config.h"

// PWM full-scale duty value derived from the board's configured resolution.
// uint32_t covers both 8-bit (GC9A01, max 255) and 13-bit (ZX2D80CE02S, max 8191).
static const uint32_t MAX_DUTY = (1u << PWM_RESOLUTION) - 1u;

// Define a PWM channel for the ESP32 Core v2 API
static const uint8_t PWM_CHANNEL = 0;

void Treadmill::begin() {
    // v2 API: setup channel, attach pin to channel, write to channel
    ledcSetup(PWM_CHANNEL, PWM_FREQ_HZ, PWM_RESOLUTION);
    ledcAttachPin(SPEED_PIN, PWM_CHANNEL);
    ledcWrite(PWM_CHANNEL, 0);

    // MCP23017 on its own I2C bus (Wire1), separate from any display/touch bus.
    Wire1.begin(MCP_I2C_SDA, MCP_I2C_SCL);
    _mcp.begin_I2C(MCP_I2C_ADDR, &Wire1);

    _mcp.pinMode(MCP_INCLINE_UP_PIN,   OUTPUT); _mcp.digitalWrite(MCP_INCLINE_UP_PIN,   LOW);
    _mcp.pinMode(MCP_INCLINE_DOWN_PIN, OUTPUT); _mcp.digitalWrite(MCP_INCLINE_DOWN_PIN, LOW);

    if (MCP_INCLINE_UP_BTN   != -1) _mcp.pinMode(MCP_INCLINE_UP_BTN,   INPUT_PULLUP);
    if (MCP_INCLINE_DOWN_BTN != -1) _mcp.pinMode(MCP_INCLINE_DOWN_BTN, INPUT_PULLUP);

    if (MCP_SAFETY_KEY_PIN != -1) {
        _mcp.pinMode(MCP_SAFETY_KEY_PIN, INPUT_PULLUP);
    }
}

void Treadmill::adjustSpeed(int32_t ticks) {
    if (_stopping) return;
    _targetSpeedMph += ticks * SPEED_STEP_MPH;
    _targetSpeedMph = constrain(_targetSpeedMph, 0.0f, MAX_SPEED_MPH);
}

void Treadmill::adjustIncline(int32_t steps) {
    if (steps == 0) return;
    int dir = (steps > 0) ? 1 : -1;
    int newLevel = constrain(_inclineLevel + dir, MIN_INCLINE_LEVEL, MAX_INCLINE_LEVEL);
    if (newLevel == _inclineLevel) return;   // already at a limit
    _inclineLevel       = newLevel;
    // Pulse the incline relay so the motor physically moves the deck.
    _inclinePulseDir    = dir;
    _inclinePulseEndMs  = millis() + INCLINE_DEBOUNCE_MS;
}

void Treadmill::toggleRunning() {
    if (_stopping) {
        // Cancel the ramp; resume at whatever speed we've ramped down to.
        _stopping = false;
    } else if (_running) {
        if (_targetSpeedMph < 0.01f) {
            // Already at zero — stop immediately.
            ledcWrite(PWM_CHANNEL, 0);
            _running = false;
        } else {
            _stopping        = true;
            _stopRampFrom    = _targetSpeedMph;
            _stopRampStartMs = millis();
        }
    } else {
        _running = true;
    }
}

void Treadmill::update() {
    bool safetyActive = (MCP_SAFETY_KEY_PIN != -1) && (_mcp.digitalRead(MCP_SAFETY_KEY_PIN) == LOW);

    if (safetyActive && !_safetyTriggered) {
        _stopAll();
        _safetyTriggered = true;
        _running  = false;
        _stopping = false;
        return;
    }

    if (_safetyTriggered) {
        _stopAll();
        if (_targetSpeedMph < 0.01f) {
            _safetyTriggered = false;
        }
        return;
    }

    if (!_running) {
        // Still service incline so on-screen / physical incline buttons work
        // while the belt is paused.
        _updateIncline();
        return;
    }

    if (_stopping) {
        uint32_t elapsed = millis() - _stopRampStartMs;
        if (elapsed >= 3000) {
            _stopAll();
            _running  = false;
            _stopping = false;
        } else {
            _targetSpeedMph = _stopRampFrom * (1.0f - (float)elapsed / 3000.0f);
            _applySpeed();
            if (_sessionRunning) {
                _elapsedSeconds = (millis() - _sessionStartMs) / 1000;
            }
        }
        return;
    }

    _updateIncline();
    _applySpeed();

    if (_speedMph > 0.01f && !_sessionRunning) {
        _sessionStartMs = millis();
        _sessionRunning = true;
    }
    if (_sessionRunning) {
        _elapsedSeconds = (millis() - _sessionStartMs) / 1000;
    }
}

void Treadmill::_applySpeed() {
    _speedMph = _targetSpeedMph;
    if (_speedMph < 0.01f) {
        ledcWrite(PWM_CHANNEL, 0);
        return;
    }
    uint32_t duty = (uint32_t)((_speedMph / MAX_SPEED_MPH) * MAX_DUTY_PCT * MAX_DUTY);
    ledcWrite(PWM_CHANNEL, duty);
}

void Treadmill::_updateIncline() {
    uint32_t now    = millis();
    bool debounceOk = (now - _lastInclineMs) > INCLINE_DEBOUNCE_MS;

    bool up   = (MCP_INCLINE_UP_BTN   != -1) && (_mcp.digitalRead(MCP_INCLINE_UP_BTN)   == LOW);
    bool down = (MCP_INCLINE_DOWN_BTN != -1) && (_mcp.digitalRead(MCP_INCLINE_DOWN_BTN) == LOW);

    // A UI-driven incline pulse still in progress drives the relay too.
    bool pulseUp   = (_inclinePulseDir > 0) && (now < _inclinePulseEndMs);
    bool pulseDown = (_inclinePulseDir < 0) && (now < _inclinePulseEndMs);
    if (_inclinePulseDir != 0 && now >= _inclinePulseEndMs) {
        _inclinePulseDir = 0;   // pulse finished
    }

    if ((up || pulseUp) && !(down || pulseDown)) {
        _mcp.digitalWrite(MCP_INCLINE_UP_PIN,   HIGH);
        _mcp.digitalWrite(MCP_INCLINE_DOWN_PIN, LOW);
        if (up && !_lastBtnUp && debounceOk) {   // physical button edge: count a level
            _inclineLevel  = min(_inclineLevel + 1, MAX_INCLINE_LEVEL);
            _lastInclineMs = now;
        }
    } else if ((down || pulseDown) && !(up || pulseUp)) {
        _mcp.digitalWrite(MCP_INCLINE_UP_PIN,   LOW);
        _mcp.digitalWrite(MCP_INCLINE_DOWN_PIN, HIGH);
        if (down && !_lastBtnDown && debounceOk) {
            _inclineLevel  = max(_inclineLevel - 1, MIN_INCLINE_LEVEL);
            _lastInclineMs = now;
        }
    } else {
        _mcp.digitalWrite(MCP_INCLINE_UP_PIN,   LOW);
        _mcp.digitalWrite(MCP_INCLINE_DOWN_PIN, LOW);
    }

    _lastBtnUp   = up;
    _lastBtnDown = down;
}

void Treadmill::_stopAll() {
    ledcWrite(PWM_CHANNEL, 0);
    _mcp.digitalWrite(MCP_INCLINE_UP_PIN,   LOW);
    _mcp.digitalWrite(MCP_INCLINE_DOWN_PIN, LOW);
    _speedMph       = 0.0f;
    _targetSpeedMph = 0.0f;  // safety reset: user must dial back to 0 to clear latch
}